Compilation error with mingw-w64 clang++ 14.0.4 from msys2

Compilation error log snippet:

```c++
[1/133] Building CXX object source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/AffineGradientSearch.cpp.obj
FAILED: source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/AffineGradientSearch.cpp.obj
ccache D:\mabs\msys64\mingw64\bin\ccache.exe  clang++ -DVVENC_SOURCE -D_WIN32_WINNT=0x0600 -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../../../include -ID:/mabs/build/vvenc-git/build-64bit -ID:/mabs/build/vvenc-git/source/Lib/vvenc/SYSTEM -ID:/mabs/build/vvenc-git/source/Lib/vvenc/. -ID:/mabs/build/vvenc-git/source/Lib/vvenc/.. -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../DecoderLib -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../EncoderLib -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/x86 -ID:/mabs/build/vvenc-git/source/Lib/vvenc/../../../thirdparty/nlohmann_json/single_include -ID:/mabs/build/vvenc-git/source/Lib/apputils/../../../include -mtune=generic -O2 -pipe -DTARGET_SIMD_X86 -DVVENC_ENABLE_THIRDPARTY_JSON -O3 -DNDEBUG -flto=thin -msse4.1 -Wall -Werror -Wno-deprecated-register -Wno-unused-const-variable -Wno-unknown-attributes -std=gnu++14 -MD -MT source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/AffineGradientSearch.cpp.obj -MF source\Lib\vvenc\CMakeFiles\vvenc.dir\__\CommonLib\AffineGradientSearch.cpp.obj.d -o source/Lib/vvenc/CMakeFiles/vvenc.dir/__/CommonLib/AffineGradientSearch.cpp.obj -c D:/mabs/build/vvenc-git/source/Lib/CommonLib/AffineGradientSearch.cpp
In file included from D:/mabs/build/vvenc-git/source/Lib/CommonLib/AffineGradientSearch.cpp:53:
In file included from D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/AffineGradientSearch.h:49:
D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/CommonDef.h:710:8: error: inline variables are a C++17 extension [-Werror,-Wc++17-extensions]
static inline unsigned long _bit_scan_reverse( long a )
       ^
D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/CommonDef.h:710:53: error: expected ')'
static inline unsigned long _bit_scan_reverse( long a )
                                                    ^
D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/CommonDef.h:710:29: note: to match this '('
static inline unsigned long _bit_scan_reverse( long a )
                            ^
D:/mabs/msys64/mingw64/lib/clang/14.0.4/include/ia32intrin.h:85:37: note: expanded from macro '_bit_scan_reverse'
#define _bit_scan_reverse(A) __bsrd((A))
                                    ^
In file included from D:/mabs/build/vvenc-git/source/Lib/CommonLib/AffineGradientSearch.cpp:53:
In file included from D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/AffineGradientSearch.h:49:
D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/CommonDef.h:710:29: error: expected expression
static inline unsigned long _bit_scan_reverse( long a )
                            ^
D:/mabs/msys64/mingw64/lib/clang/14.0.4/include/ia32intrin.h:85:40: note: expanded from macro '_bit_scan_reverse'
#define _bit_scan_reverse(A) __bsrd((A))
                                       ^
In file included from D:/mabs/build/vvenc-git/source/Lib/CommonLib/AffineGradientSearch.cpp:53:
In file included from D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/AffineGradientSearch.h:49:
D:/mabs/build/vvenc-git/source/Lib/vvenc/../CommonLib/CommonDef.h:710:56: error: expected ';' after top level declarator
static inline unsigned long _bit_scan_reverse( long a )
                                                       ^
                                                       ;
```

Additional errors do exist, but mainly due to CommonDef.h becoming malformed

Main suggestion is to either remove the leading underscore (since leading underscore names are generally reserved `reserved names include all external identifiers (global functions and variables) that begin with an underscore (‘_’)` <https://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html>, and although this isn't an external identifier, it would still be a "good idea" to prevent issues with macros and functions declared and defined by the compiler and libs)

--- a/source/Lib/CommonLib/CommonDef.h
+++ b/source/Lib/CommonLib/CommonDef.h
@@ -705,6 +705,7 @@ const char* read_x86_extension(const std::string &extStrId);
 template <typename ValueType> inline ValueType leftShiftU  (const ValueType value, const unsigned shift) { return value << shift; }
 template <typename ValueType> inline ValueType rightShiftU (const ValueType value, const unsigned shift) { return value >> shift; }
 
+#if ENABLE_SIMD_LOG2
 #if defined( _WIN32 ) && defined( TARGET_SIMD_X86 )
 static inline unsigned int bit_scan_reverse( int a )
 {
@@ -724,8 +725,6 @@ static inline unsigned int bit_scan_reverse( int a )
   return __builtin_clz( a ) ^ ( 8 * sizeof( a ) - 1 );
 }
 #endif
-
-#if ENABLE_SIMD_LOG2
 static inline int floorLog2( int val )
 {
   return bit_scan_reverse( val );

--- a/source/Lib/CommonLib/CommonDef.h
+++ b/source/Lib/CommonLib/CommonDef.h
@@ -705,7 +705,6 @@ const char* read_x86_extension(const std::string &extStrId);
 template <typename ValueType> inline ValueType leftShiftU  (const ValueType value, const unsigned shift) { return value << shift; }
 template <typename ValueType> inline ValueType rightShiftU (const ValueType value, const unsigned shift) { return value >> shift; }
 
-#if ENABLE_SIMD_LOG2
 #if defined( _WIN32 ) && defined( TARGET_SIMD_X86 )
 static inline unsigned int bit_scan_reverse( int a )
 {
@@ -725,6 +724,7 @@ static inline unsigned int bit_scan_reverse( int a )
   return __builtin_clz( a ) ^ ( 8 * sizeof( a ) - 1 );
 }
 #endif
+#if ENABLE_SIMD_LOG2
 static inline int floorLog2( int val )
 {
   return bit_scan_reverse( val );

--- a/source/Lib/CommonLib/CommonDef.h
+++ b/source/Lib/CommonLib/CommonDef.h
@@ -705,21 +705,31 @@ const char* read_x86_extension(const std::string &extStrId);
 template <typename ValueType> inline ValueType leftShiftU  (const ValueType value, const unsigned shift) { return value << shift; }
 template <typename ValueType> inline ValueType rightShiftU (const ValueType value, const unsigned shift) { return value >> shift; }
 
-#ifdef TARGET_SIMD_X86
-#ifdef _WIN32
-static inline unsigned long _bit_scan_reverse( long a )
+#if defined( _WIN32 ) && defined( TARGET_SIMD_X86 )
+static inline unsigned int bit_scan_reverse( int a )
 {
   unsigned long idx = 0;
   _BitScanReverse( &idx, a );
   return idx;
 }
+// disabled because it requires x86intrin.h which conflicts with simd-everywhere
+// #elif defined( __GNUC__ ) && defined( TARGET_SIMD_X86 ) && !defined( REAL_TARGET_WASM )
+// static inline unsigned int bit_scan_reverse( int a )
+// {
+//   return _bit_scan_reverse( a );
+// }
+#elif defined( __GNUC__ )
+static inline unsigned int bit_scan_reverse( int a )
+{
+  return __builtin_clz( a ) ^ ( 8 * sizeof( a ) - 1 );
+}
 #endif
 
-static inline int floorLog2( uint32_t x )
+#if ENABLE_SIMD_LOG2
+static inline int floorLog2( int val )
 {
-  return ( int ) _bit_scan_reverse( x );
+  return bit_scan_reverse( val );
 }
-
 #else
 static inline int floorLog2(uint32_t x)
 {

--- a/source/Lib/CommonLib/DepQuant.cpp
+++ b/source/Lib/CommonLib/DepQuant.cpp
@@ -649,25 +649,6 @@ namespace DQIntern
     int64_t           m_DistOrgFact;
   };
 
-  inline int ceil_log2(uint64_t x)
-  {
-#ifdef TARGET_SIMD_X86
-    uint64_t y = _bit_scan_reverse( x ) + ( ( ( x & ( x - 1 ) ) == 0 ) ? 0 : 1 );
-#else
-    static const uint64_t t[6] = { 0xFFFFFFFF00000000ull, 0x00000000FFFF0000ull, 0x000000000000FF00ull, 0x00000000000000F0ull, 0x000000000000000Cull, 0x0000000000000002ull };
-    int y = (((x & (x - 1)) == 0) ? 0 : 1);
-    int j = 32;
-    for( int i = 0; i < 6; i++)
-    {
-      int k = (((x & t[i]) == 0) ? 0 : j);
-      y += k;
-      x >>= k;
-      j >>= 1;
-    }
-#endif
-
-    return y;
-  }
   void Quantizer::initQuantBlock(const TransformUnit& tu, const ComponentID compID, const QpParam& cQP, const double lambda, int gValue = -1)
   {
     CHECKD( lambda <= 0.0, "Lambda must be greater than 0" );
@@ -703,7 +684,7 @@ namespace DQIntern
     const double  qScale2       = double( qScale * qScale );
     const double  nomDistFactor = ( nomDShift < 0 ? 1.0/(double(int64_t(1)<<(-nomDShift))*qScale2*lambda) : double(int64_t(1)<<nomDShift)/(qScale2*lambda) );
     const int64_t pow2dfShift   = (int64_t)( nomDistFactor * qScale2 ) + 1;
-    const int     dfShift       = ceil_log2( pow2dfShift );
+    const int     dfShift       = ceilLog2( pow2dfShift );
     m_DistShift                 = 62 + m_QShift - 2*maxLog2TrDynamicRange - dfShift;
     m_DistAdd                   = (int64_t(1) << m_DistShift) >> 1;
     m_DistStepAdd               = ((m_DistShift+m_QShift)>=64 ? (int64_t)( nomDistFactor * pow(2,m_DistShift+m_QShift) + .5 ) : (int64_t)( nomDistFactor * double(int64_t(1)<<(m_DistShift+m_QShift)) + .5 ));

--- a/source/Lib/CommonLib/TypeDef.h
+++ b/source/Lib/CommonLib/TypeDef.h
@@ -144,6 +144,7 @@ namespace vvenc {
 #define ENABLE_SIMD_OPT_MCTF                            ( 1 && ENABLE_SIMD_OPT )                            ///< SIMD optimization for MCTF
 #define ENABLE_SIMD_TRAFO                               ( 1 && ENABLE_SIMD_OPT )                            ///< SIMD optimization for Transformation
 #define ENABLE_SIMD_OPT_QUANT                           ( 1 && ENABLE_SIMD_OPT )                            ///< SIMD optimization for Quantization
+#define ENABLE_SIMD_LOG2                                ( 1 && ENABLE_SIMD_OPT )                            ///< use SIMD intrisic to calculate log2
 
 #if ENABLE_SIMD_OPT_BUFFER
 #define ENABLE_SIMD_OPT_BCW                               1                                                 ///< SIMD optimization for GBi

--- a/source/Lib/CommonLib/x86/InterPredX86.h
+++ b/source/Lib/CommonLib/x86/InterPredX86.h
@@ -70,7 +70,7 @@ namespace vvenc {
 
 static inline int rightShiftMSB(int numer, int denom)
 {
-  int shiftIdx = _bit_scan_reverse(denom);
+  int shiftIdx = bit_scan_reverse(denom);
   return (numer >> shiftIdx);
 }
   

